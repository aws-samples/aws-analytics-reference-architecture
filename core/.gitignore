# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/package.json
!/LICENSE
!/.npmignore
logs
*.log
npm-debug.log*
yarn-debug.log*
yarn-error.log*
lerna-debug.log*
report.[0-9]*.[0-9]*.[0-9]*.[0-9]*.json
pids
*.pid
*.seed
*.pid.lock
lib-cov
coverage
*.lcov
.nyc_output
build/Release
node_modules/
jspm_packages/
*.tsbuildinfo
.eslintcache
*.tgz
.yarn-integrity
.cache
!/.projenrc.js
/test-reports/
junit.xml
/coverage/
/dist/changelog.md
/dist/version.txt
!/.mergify.yml
!/test/
!/src/
/lib
/dist/
!/tsconfig.jest.json
!/.eslintrc.json
!/tsconfig.eslint.json
.jsii
tsconfig.json
!/API.md
build
!/.github/workflows/release.yml
