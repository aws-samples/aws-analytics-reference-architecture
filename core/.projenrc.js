// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

const { basename, join, dirname, relative } = require('path');
const glob = require('glob');


const { awscdk } = require('projen');

const project = new awscdk.AwsCdkConstructLibrary({

  authorName: 'Amazon Web Services',
  authorUrl: 'https://aws.amazon.com',
  authorOrganization: true,
  homepage: 'https://aws-samples.github.io/aws-analytics-reference-architecture/',
  copyrightPeriod: `2021-${new Date().getFullYear()}`,
  copyrightOwner: 'Amazon.com, Inc. or its affiliates. All Rights Reserved.',
  keywords: [
    'aws',
    'constructs',
    'cdk',
    'analytics',
  ],

  cdkVersion: '1.155.0',
  defaultReleaseBranch: 'main',
  license: 'MIT-0',
  name: 'aws-analytics-reference-architecture',
  repositoryUrl: 'https://github.com/aws-samples/aws-analytics-reference-architecture.git',
  repositoryDirectory: 'core',
  workflow: false,
  buildWorkflow: false,
  release: true,
  depsUpgrade: true,
  stale: false,
  pullRequestTemplate: false,
  cdkVersionPinning: true,
  githubOptions: {
    pullRequestLint: false,
  },
  //workflowContainerImage: 'jsii/superchain:1-buster-slim',

  cdkDependencies: [
    '@aws-cdk/assertions',
    '@aws-cdk/aws-athena',
    '@aws-cdk/aws-autoscaling',
    '@aws-cdk/aws-ec2',
    '@aws-cdk/aws-emrcontainers',
    '@aws-cdk/aws-eks',
    '@aws-cdk/aws-events',
    '@aws-cdk/aws-events-targets',
    '@aws-cdk/aws-glue',
    '@aws-cdk/aws-iam',
    '@aws-cdk/aws-kinesis',
    '@aws-cdk/aws-kinesisfirehose',
    '@aws-cdk/aws-kinesisfirehose-destinations',
    '@aws-cdk/aws-lambda',
    '@aws-cdk/aws-lambda-python',
    '@aws-cdk/aws-logs',
    '@aws-cdk/aws-redshift',
    '@aws-cdk/aws-s3',
    '@aws-cdk/aws-s3-assets',
    '@aws-cdk/aws-s3-deployment',
    '@aws-cdk/aws-secretsmanager',
    '@aws-cdk/aws-stepfunctions',
    '@aws-cdk/aws-stepfunctions-tasks',
    '@aws-cdk/core',
    '@aws-cdk/custom-resources',
    '@aws-cdk/lambda-layer-awscli',
    '@aws-cdk/aws-emr',
    '@aws-cdk/aws-kms',
    '@aws-cdk/aws-lakeformation',
  ],

  deps: [
    '@exodus/schemasafe',
  ],

  devDeps: [
    '@types/js-yaml',
    '@types/jest',
    'esbuild',
    'aws-cdk@1.155.0',
    'cdk-nag@^1.0.0',
    'cdk-assets@1.155.0',
    'jest-runner-groups',
    'promptly',
    'proxy-agent',
    'glob',
  ],

  jestOptions: {
    jestConfig: {
      runner: 'groups',
    },
  },

  bundledDeps: [
    'js-yaml',
    'uuid',
    'aws-sdk',
    '@exodus/schemasafe',
  ],

  python: {
    distName: 'aws_analytics_reference_architecture',
    module: 'aws_analytics_reference_architecture',
  },

  tsconfig: {
    compilerOptions: {
      resolveJsonModule: true,
      esModuleInterop: true,
    },
    include: ['src/**/*.json'],
  },

  stability: 'experimental',

});

project.testTask.reset('jest --group=unit');

project.testTask.spawn('eslint');

project.addTask('test:best-practice', {
  exec: 'jest --group=best-practice',
});

project.addTask('test:unit', {
  exec: 'jest --group=unit',
});

project.addTask('test:integ', {
  exec: 'jest --group=integ',
});

project.addTask('test:integ/data-lake', {
  exec: 'jest --group=integ/data-lake',
});

project.addTask('test:integ/redshift', {
  exec: 'jest --group=integ/redshift',
});

const testDeploy = project.addTask('test:deploy', {
  exec: 'cdk --version && cdk deploy --app=./lib/integ.default.js',
});

testDeploy.prependExec('npx projen build');

project.packageTask.spawn(project.tasks.tryFind('package-all'));

project.addTask('test:destroy', {
  exec: 'cdk destroy --app=./lib/integ.default.js',
});


project.addDevDeps('glob');

/**
 * Task to copy `resources` directories from `src` to `lib`
 */

const copyResourcesToLibTask = project.addTask('copy-resources', {
  description: 'Copy all resources directories from src to lib',
});

for (const from of glob.sync('src/**/resources')) {
  const to = dirname(from.replace('src', 'lib'));
  const cpCommand = `rsync -avr --exclude '*.ts' --exclude '*.js' ${from} ${to}`;
  copyResourcesToLibTask.exec(cpCommand);
}

/**
 * Task to pip install all Python Lambda functions in lib folder
 */

const pipInstallTask = project.addTask('pip-install', {
  description: 'pip install all folders in lib that has requirements.txt',
});

for (const dirPath of findAllPythonLambdaDir('src')) {
  // Assume that all folders with 'requirements.txt' have been copied to lib
  // by the task 'copy-resources'
  const dirPathInLib = dirPath.replace('src', 'lib');
  const target = dirname(dirPathInLib);
  const pipCmd = `pip3 install -r ${dirPathInLib} --target ${target} --upgrade`;

  pipInstallTask.exec(pipCmd);
}

/**
 * Task to build java lambda jar with gradle
 */
const gradleBuildTask = project.addTask('gradle-build', {
  description: './gradlew shadowJar all folders in lib that has requirements.txt',
});

for (const gradlePath of findAllGradleLambdaDir('src')) {
  console.log('loop over gradle dir');
  const dirPath = dirname(gradlePath);
  const gradleCmd = `cd ${dirPath} && ./gradlew shadowJar && cp build/libs/*.jar ./ 2> /dev/null`;

  gradleBuildTask.exec(gradleCmd);
}

/**
 * Run `copy-resources` and `pip-install` as part of compile
 */
project.compileTask.exec('npx projen gradle-build');
project.compileTask.exec('npx projen copy-resources');
project.compileTask.exec('npx projen pip-install');

/**
 * Find all directory that has a Python package.
 * Assume that they have requirements.txt
 *
 * @param rootDir Root directory to begin finding
 * @returns Array of directory paths
 */
function findAllPythonLambdaDir(rootDir) {

  return glob.sync(`${rootDir}/**/requirements.txt`).map((pathWithReq) => {
    return pathWithReq;
  });
}

/**
 * Find all directory that has a gradle package.
 * Assume that they have build.gradle
 *
 * @param rootDir Root directory to begin finding
 * @returns Array of directory paths
 */
function findAllGradleLambdaDir(rootDir) {
  console.log('findAllGradleLambdaDir');

  return glob.sync(`${rootDir}/**/build.gradle`).map((pathWithGradle) => {
    return pathWithGradle;
  });
}

project.synth();
