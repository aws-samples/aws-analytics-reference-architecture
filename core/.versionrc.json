{
  "packageFiles": [
    {
      "filename": "/Users/gromav/Local/tmp/aws-analytics-reference-architecture/core/package.json",
      "type": "json"
    }
  ],
  "bumpFiles": [
    {
      "filename": "/Users/gromav/Local/tmp/aws-analytics-reference-architecture/core/package.json",
      "type": "json"
    }
  ],
  "commitAll": false,
  "infile": "/Users/gromav/Local/tmp/aws-analytics-reference-architecture/core/dist/changelog.md",
  "header": "",
  "skip": {
    "commit": true,
    "tag": true,
    "bump": false
  }
}