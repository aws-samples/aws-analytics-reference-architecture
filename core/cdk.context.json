{
  "availability-zones:account=214783019211:region=eu-west-1": [
    "eu-west-1a",
    "eu-west-1b",
    "eu-west-1c"
  ]
}
