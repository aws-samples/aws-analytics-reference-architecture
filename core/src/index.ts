// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

export { ExampleProps, Example } from './example';
export { DataLakeStorageProps, DataLakeStorage } from './data-lake-storage';
export { SynchronousCrawlerProps, SynchronousCrawler } from './synchronous-crawler';
export { DatasetProps, Dataset, PartitionedDatasetProps, PartitionedDataset } from './datasets';
export { DataGeneratorProps, DataGenerator, BatchReplayerProps, BatchReplayer } from './data-generator';
export { SynchronousAthenaQueryProps, SynchronousAthenaQuery } from './synchronous-athena-query';
export { SingletonBucket } from './singleton-bucket';
export { Ec2SsmRole } from './ec2-ssm-role';
export { DataLakeCatalog } from './data-lake-catalog';
export { DataLakeExporter, DataLakeExporterProps } from './data-lake-exporter';
export { AthenaDefaultSetup } from './athena-default-setup';
export { SingletonGlueDefaultRole } from './singleton-glue-default-role';
export { EmrEksClusterProps, EmrEksCluster, EmrEksNodegroupOptions, EmrEksNodegroup, EmrVirtualClusterProps } from './emr-eks-data-platform';
export { FlywayRunner, FlywayRunnerProps } from './db-schema-manager';
