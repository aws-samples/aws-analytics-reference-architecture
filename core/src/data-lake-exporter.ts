import { Database, Table } from '@aws-cdk/aws-glue';
import { Role, ServicePrincipal, PolicyStatement } from '@aws-cdk/aws-iam';
import { Stream } from '@aws-cdk/aws-kinesis';
import { CfnDeliveryStream } from '@aws-cdk/aws-kinesisfirehose';
import { LogGroup, RetentionDays, LogStream } from '@aws-cdk/aws-logs';
import { Bucket, Location } from '@aws-cdk/aws-s3';
import { Construct, Aws, RemovalPolicy, Stack } from '@aws-cdk/core';


/**
 * The properties for DataLakeExporter Construct.
 */
export interface DataLakeExporterProps {
  /**
   * Sink must be an Amazon S3 Location composed of a bucket and a key.
   */
  readonly sinkLocation: Location;
  /**
   * Source must be an Amazon Kinesis Data Stream.
   */
  readonly sourceKinesisDataStream: Stream;
  /**
   * Source AWS Glue Database containing the schema of the stream.
   */
  readonly sourceGlueDatabase: Database;
  /**
   * Source AWS Glue Table containing the schema of the stream.
   */
  readonly sourceGlueTable: Table;
  /**
   * Delivery interval in seconds. The frequency of the data delivery is defined by this interval.
   * @default - Set to 900 seconds
   */
  readonly deliveryInterval?: Number;
  /**
   * Maximum delivery size in MB. The frequency of the data delivery is defined by this maximum delivery size.
   * @default - Set to 128 MB
   */
  readonly deliverySize?: Number;
}


/**
 * DataLakeExporter Construct to export data from a stream to the data lake.
 * Source can be an Amazon Kinesis Data Stream.
 * Target can be an Amazon S3 bucket.
 */
export class DataLakeExporter extends Construct {

  /**
   * Constructs a new instance of the DataLakeExporter class
   * @param {Construct} scope the Scope of the AWS CDK Construct
   * @param {string} id the ID of the AWS CDK Construct
   * @param {DataLakeExporterProps} props the DataLakeExporter [properties]{@link DataLakeExporterProps}
   * @access public
   */
  public readonly cfnIngestionStream: CfnDeliveryStream;

  constructor(scope: Construct, id: string, props: DataLakeExporterProps) {
    super(scope, id);

    if ( props.deliverySize || 128 > 128 ) { throw 'deliverySize cannot be more than 128MB';}
    if ( props.deliveryInterval || 900 > 900 ) { throw 'deliveryInterval cannot be more than 900s';}

    const stack = Stack.of(this);

    // Get the Bucket from Amazon S3 Location sink
    const sinkBucket = Bucket.fromBucketName(this, 'sinkBucket', props.sinkLocation.bucketName);

    // Create log group for storing Amazon Kinesis Firehose logs.
    const logGroup = new LogGroup(this, 'dataLakeExporterLogGroup', {
      logGroupName: '/data-lake-exporter/',
      removalPolicy: RemovalPolicy.DESTROY,
      retention: RetentionDays.ONE_DAY,
    });

    // Create the Kinesis Firehose log stream.
    const firehoseLogStream = new LogStream(this, 'dataLakeExporterLogStream', {
      logGroup: logGroup,
      logStreamName: 'firehose-stream',
      removalPolicy: RemovalPolicy.DESTROY,
    });

    // Create an Amazon IAM Role used by Amazon Kinesis Firehose delivery stream
    const role = new Role(this, 'dataLakeExporterRole', {
      assumedBy: new ServicePrincipal('firehose.amazonaws.com'),
    });

    // add policy to access AWS Glue Database and Table for format conversion
    role.addToPolicy(
      new PolicyStatement({
        resources: [
          props.sourceGlueTable.tableArn,
          props.sourceGlueDatabase.catalogArn,
          props.sourceGlueDatabase.databaseArn,
        ],
        actions: [
          'glue:GetTable',
          'glue:GetTableVersion',
          'glue:GetTableVersions',
        ],
      }),
    );

    // add policy to access source stream
    role.addToPolicy(
      new PolicyStatement({
        resources: [
          props.sourceKinesisDataStream.streamArn,
        ],
        actions: [
          'kinesis:DescribeStream',
          'kinesis:GetShardIterator',
          'kinesis:GetRecords',
          'kinesis:ListShards',
        ],
      }),
    );

    // add policy to write data in Amazon S3 Location
    role.addToPolicy(
      new PolicyStatement({
        resources: [
          stack.formatArn({
            account: '',
            region: '',
            service: 's3',
            resource: props.sinkLocation.bucketName,
            resourceName: props.sinkLocation.objectKey,
          }),
          stack.formatArn({
            account: '',
            region: '',
            service: 's3',
            resource: props.sinkLocation.bucketName,
            resourceName: `${props.sinkLocation.objectKey}/*`,
          }),
          stack.formatArn({
            account: '',
            region: '',
            service: 's3',
            resource: props.sinkLocation.bucketName,
          }),
        ],
        actions: [
          's3:AbortMultipartUpload',
          's3:GetBucketLocation',
          's3:GetObject',
          's3:ListBucket',
          's3:ListBucketMultipartUploads',
          's3:PutObject',
        ],
      }),
    );

    // add policy to write logs to CloudWatch logs
    role.addToPolicy(
      new PolicyStatement({
        resources: [
          `${logGroup.logGroupArn}:log-stream:${firehoseLogStream.logStreamName}`,
        ],
        actions: [
          'logs:PutLogEvents',
        ],
      }),
    );

    // TODO add policy for KMS managed?

    /*     this.ingestionStream = new DeliveryStream(this, 'dataLakeExporter', {
      sourceStream: props.sourceKinesisDataStream,
      destinations: [new S3Bucket(sinkBucket,{
        dataOutputPrefix: props.sinkLocation.objectKey,
        errorOutputPrefix: `${props.sinkLocation.objectKey}-error`,
        logGroup: logGroup,
        compression: Compression.SNAPPY,
        bufferingInterval: Duration.seconds(props.deliveryInterval || 900),
        bufferingSize: Size.mebibytes(props.deliverySize || 128),
      })],
      encryption: StreamEncryption.AWS_OWNED,
    }) */

    // Create the Delivery stream from Cfn because L2 Construct doesn't support conversion to parquet and custom partitioning
    this.cfnIngestionStream = new CfnDeliveryStream(this, 'dataLakeExporter', {
      deliveryStreamType: 'KinesisStreamAsSource',
      deliveryStreamEncryptionConfigurationInput: {
        keyType: 'AWS_OWNED_CMK',
      },
      extendedS3DestinationConfiguration: {
        bucketArn: sinkBucket.bucketArn,
        bufferingHints: {
          intervalInSeconds: props.deliveryInterval || 900,
          sizeInMBs: props.deliverySize || 128,
        },
        cloudWatchLoggingOptions: {
          logGroupName: logGroup.logGroupName,
          logStreamName: firehoseLogStream.logStreamName,
        },
        roleArn: role.roleArn,
        errorOutputPrefix: `${props.sinkLocation.objectKey}-error`,
        prefix: props.sinkLocation.objectKey,
        compressionFormat: 'UNCOMPRESSED',
        s3BackupMode: 'Disabled',
        dataFormatConversionConfiguration: {
          enabled: true,
          inputFormatConfiguration: {
            deserializer: {
              openXJsonSerDe: {},
            },
          },
          outputFormatConfiguration: {
            serializer: {
              parquetSerDe: {},
            },
          },
          schemaConfiguration: {
            roleArn: role.roleArn,
            catalogId: Aws.ACCOUNT_ID,
            region: Aws.REGION,
            databaseName: props.sourceGlueDatabase.databaseName,
            tableName: props.sourceGlueTable.tableName,
          },
        },
      },
      kinesisStreamSourceConfiguration: {
        kinesisStreamArn: '',
        roleArn: role.roleArn,
      },
    });
  }
}
