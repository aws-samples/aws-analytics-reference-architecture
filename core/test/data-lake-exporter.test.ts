import { Stream } from '@aws-cdk/aws-kinesis';
import { Stack } from '@aws-cdk/core';
import { DataLakeExporter } from '../src/data-lake-exporter';
import '@aws-cdk/assert/jest';
import { Database, DataFormat, Table } from '@aws-cdk/aws-glue';

test('dataLakeExporter', () => {
  const dataLakeExporterStack = new Stack();

  const stream = new Stream(dataLakeExporterStack, 'testStream');
  const db = new Database(dataLakeExporterStack, 'testDB', { databaseName: 'test_db' });
  const table = new Table(dataLakeExporterStack, 'testTable', {
    database: db,
    tableName: 'test_table',
    dataFormat: DataFormat.JSON,
    columns: [
      {
        name: 'a',
        type: {
          isPrimitive: true,
          inputString: 'STRING',
        },
      },
    ],
  });

  new DataLakeExporter(dataLakeExporterStack, 'testExporter', {
    sinkLocation: {
      bucketName: 'test',
      objectKey: 'test',
    },
    sourceKinesisDataStream: stream,
    sourceGlueDatabase: db,
    sourceGlueTable: table,
  });

  // TODO: add tests (refer to data-lake-storage.test.ts)

});
