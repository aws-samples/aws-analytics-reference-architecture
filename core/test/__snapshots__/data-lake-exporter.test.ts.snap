// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dataLakeExporter 1`] = `
Object {
  "Conditions": Object {
    "AwsCdkKinesisEncryptedStreamsUnsupportedRegions": Object {
      "Fn::Or": Array [
        Object {
          "Fn::Equals": Array [
            Object {
              "Ref": "AWS::Region",
            },
            "cn-north-1",
          ],
        },
        Object {
          "Fn::Equals": Array [
            Object {
              "Ref": "AWS::Region",
            },
            "cn-northwest-1",
          ],
        },
      ],
    },
  },
  "Resources": Object {
    "testDB74E9B281": Object {
      "Properties": Object {
        "CatalogId": Object {
          "Ref": "AWS::AccountId",
        },
        "DatabaseInput": Object {
          "Name": "test_db",
        },
      },
      "Type": "AWS::Glue::Database",
    },
    "testExporterdataLakeExporter51BD8EA9": Object {
      "Properties": Object {
        "DeliveryStreamEncryptionConfigurationInput": Object {
          "KeyType": "AWS_OWNED_CMK",
        },
        "DeliveryStreamType": "KinesisStreamAsSource",
        "ExtendedS3DestinationConfiguration": Object {
          "BucketARN": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":s3:::test",
              ],
            ],
          },
          "BufferingHints": Object {
            "IntervalInSeconds": 900,
            "SizeInMBs": 128,
          },
          "CloudWatchLoggingOptions": Object {
            "LogGroupName": Object {
              "Ref": "testExporterdataLakeExporterLogGroup610F71C7",
            },
            "LogStreamName": Object {
              "Ref": "testExporterdataLakeExporterLogStreamF2BE8809",
            },
          },
          "CompressionFormat": "UNCOMPRESSED",
          "DataFormatConversionConfiguration": Object {
            "Enabled": true,
            "InputFormatConfiguration": Object {
              "Deserializer": Object {
                "OpenXJsonSerDe": Object {},
              },
            },
            "OutputFormatConfiguration": Object {
              "Serializer": Object {
                "ParquetSerDe": Object {},
              },
            },
            "SchemaConfiguration": Object {
              "CatalogId": Object {
                "Ref": "AWS::AccountId",
              },
              "DatabaseName": Object {
                "Ref": "testDB74E9B281",
              },
              "Region": Object {
                "Ref": "AWS::Region",
              },
              "RoleARN": Object {
                "Fn::GetAtt": Array [
                  "testExporterdataLakeExporterRole1BEE617B",
                  "Arn",
                ],
              },
              "TableName": Object {
                "Ref": "testTable12C06BD8",
              },
            },
          },
          "ErrorOutputPrefix": "test-error",
          "Prefix": "test",
          "RoleARN": Object {
            "Fn::GetAtt": Array [
              "testExporterdataLakeExporterRole1BEE617B",
              "Arn",
            ],
          },
          "S3BackupMode": "Disabled",
        },
        "KinesisStreamSourceConfiguration": Object {
          "KinesisStreamARN": "",
          "RoleARN": Object {
            "Fn::GetAtt": Array [
              "testExporterdataLakeExporterRole1BEE617B",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::KinesisFirehose::DeliveryStream",
    },
    "testExporterdataLakeExporterLogGroup610F71C7": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "LogGroupName": "/data-lake-exporter/",
        "RetentionInDays": 1,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "testExporterdataLakeExporterLogStreamF2BE8809": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "LogGroupName": Object {
          "Ref": "testExporterdataLakeExporterLogGroup610F71C7",
        },
        "LogStreamName": "firehose-stream",
      },
      "Type": "AWS::Logs::LogStream",
      "UpdateReplacePolicy": "Delete",
    },
    "testExporterdataLakeExporterRole1BEE617B": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "firehose.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "testExporterdataLakeExporterRoleDefaultPolicy3EEE6D56": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "glue:GetTable",
                "glue:GetTableVersion",
                "glue:GetTableVersions",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":glue:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":table/",
                      Object {
                        "Ref": "testDB74E9B281",
                      },
                      "/",
                      Object {
                        "Ref": "testTable12C06BD8",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":glue:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":catalog",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":glue:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":database/",
                      Object {
                        "Ref": "testDB74E9B281",
                      },
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "kinesis:DescribeStream",
                "kinesis:GetShardIterator",
                "kinesis:GetRecords",
                "kinesis:ListShards",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "testStream8BCA7523",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": Array [
                "s3:AbortMultipartUpload",
                "s3:GetBucketLocation",
                "s3:GetObject",
                "s3:ListBucket",
                "s3:ListBucketMultipartUploads",
                "s3:PutObject",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::test/test",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::test/test/*",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::test",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "logs:PutLogEvents",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "testExporterdataLakeExporterLogGroup610F71C7",
                        "Arn",
                      ],
                    },
                    ":log-stream:",
                    Object {
                      "Ref": "testExporterdataLakeExporterLogStreamF2BE8809",
                    },
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testExporterdataLakeExporterRoleDefaultPolicy3EEE6D56",
        "Roles": Array [
          Object {
            "Ref": "testExporterdataLakeExporterRole1BEE617B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testStream8BCA7523": Object {
      "Properties": Object {
        "RetentionPeriodHours": 24,
        "ShardCount": 1,
        "StreamEncryption": Object {
          "Fn::If": Array [
            "AwsCdkKinesisEncryptedStreamsUnsupportedRegions",
            Object {
              "Ref": "AWS::NoValue",
            },
            Object {
              "EncryptionType": "KMS",
              "KeyId": "alias/aws/kinesis",
            },
          ],
        },
      },
      "Type": "AWS::Kinesis::Stream",
    },
    "testTable12C06BD8": Object {
      "Properties": Object {
        "CatalogId": Object {
          "Ref": "AWS::AccountId",
        },
        "DatabaseName": Object {
          "Ref": "testDB74E9B281",
        },
        "TableInput": Object {
          "Description": "test_table generated by CDK",
          "Name": "test_table",
          "Parameters": Object {
            "classification": "json",
            "has_encrypted_data": false,
          },
          "StorageDescriptor": Object {
            "Columns": Array [
              Object {
                "Name": "a",
                "Type": "STRING",
              },
            ],
            "Compressed": false,
            "InputFormat": "org.apache.hadoop.mapred.TextInputFormat",
            "Location": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "s3://",
                  Object {
                    "Ref": "testTableBucket030117E6",
                  },
                  "/",
                ],
              ],
            },
            "OutputFormat": "org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat",
            "SerdeInfo": Object {
              "SerializationLibrary": "org.openx.data.jsonserde.JsonSerDe",
            },
            "StoredAsSubDirectories": false,
          },
          "TableType": "EXTERNAL_TABLE",
        },
      },
      "Type": "AWS::Glue::Table",
    },
    "testTableBucket030117E6": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
